### PR TITLE
Add Triangle

### PR DESCRIPTION
**Pull request by Miou-zora**
render

### CODE REVIEW

File: src/Triangle.zig [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+pub const Triangle = struct {
+    const Self = @This();
+
+    va: Vertex,
+    vb: Vertex,
+    vc: Vertex,
+
+    pub fn hits(self: *const Self, ray: Ray) HitRecord {
+        const a = self.va.position;
```
Comment (Miou-zora) [inline src/Triangle.zig:15]:
```suggestion
        // TODO: add bvh
        const a = self.va.position;
```

File: src/Triangle.zig [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        const material: Material = .{
+            .color = colorRGB,
+            .reflective = 0.5,
+            .specular = 0,
+        };
+        return HitRecord{
+            // .hit = true,
+            // .t = t,
+            // .hit_point = hit_point,
+            // .normal = normal,
```
Comment (huntears) [inline src/Triangle.zig:58]: No need to keep this

File: src/Vertex.zig [hunk 3]
```diff
@@ -0,0 +1,12 @@
+const Pt3 = @import("Pt3.zig").Pt3;
+const ColorRGB = @import("ColorRGB.zig").ColorRGB;
+const Vec3 = @import("Vec3.zig").Vec3;
+
+pub const Vertex = struct {
+    const Self = @This();
+
+    position: Pt3,
+    color: ColorRGB,
+    // normal: Vec3,
```
Comment (huntears) [inline src/Vertex.zig:10]: Why have a normal with a vertex?

File: src/main.zig [hunk 4]
```diff
@@ -15,6 +15,7 @@ const Scene = @import("Scene.zig");
 const ColorRGB = @import("ColorRGB.zig").ColorRGB;
 const Material = @import("Material.zig").Material;
 const Config = @import("Config.zig").Config;
+// const Triangle = @import("Triangle.zig").Triangle;
```
Comment (huntears) [inline src/main.zig:18]: Please remove

File: src/Triangle.zig [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+pub const Triangle = struct {
+    const Self = @This();
+
+    va: Vertex,
+    vb: Vertex,
+    vc: Vertex,
+
+    pub fn hits(self: *const Self, ray: Ray) HitRecord {
+        // TODO: add bvh
```
Comment (Miou-zora) [inline src/Triangle.zig:15]:
```suggestion
        // TODO: add bvh + compute this only one time
```

File: src/main.zig [hunk 6]
```diff
@@ -25,7 +26,7 @@ pub fn compute_lighting(intersection: Vec3, normal: Vec3, scene: *Scene.Scene, r
             .point_light => |item| {
                 const L = intersection.to(item.position).normalized();
                 t_max = intersection.to(item.position).length();
-                const closest_hit = find_closest_intersection(scene, Ray{ .direction = L, .origin = intersection.addVec3(normal.mulf32(0.0001)) }, 0.0001, t_max);
+                const closest_hit = find_closest_intersection(scene, Ray{ .direction = L, .origin = intersection.addVec3(normal.mulf32(0.00001)) }, 0.00001, t_max);
```
Comment (huntears) [inline src/main.zig:29]: You really need that kind of magic small number?

File: src/main.zig [hunk 7]
```diff
@@ -142,6 +143,13 @@ fn find_closest_intersection(scene: *Scene.Scene, ray: Ray, t_min: f32, t_max: f
                     }
                 }
             },
+            .triangle => |item| {
+                // std.debug.print("{}\n", .{item});
```
Comment (huntears) [inline src/main.zig:147]: Please remove

File: src/main.zig [hunk 8]
```diff
@@ -223,6 +231,22 @@ pub fn main() !void {
...
+    //         .color = ColorRGB{ .r = 0, .g = 0, .b = 255 },
+    //         .position = Pt3{ .x = -2, .y = 3, .z = 3.5 },
+    //     },
+    //     .vc = .{
+    //         .color = ColorRGB{ .r = 0, .g = 255, .b = 0 },
+    //         .position = Pt3{ .x = 0.5, .y = -0.5, .z = 4 },
+    //     },
+    // } };
+    // try scene.objects.append(triangle);
+
```
Comment (huntears) [inline src/main.zig:249]: Please remove

**Review by huntears — CHANGES_REQUESTED** [review]
Looks cool :3

File: src/Triangle.zig [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        const u = bSuba.cross(hit_point.subVec3(a)).dot(normal);
+        const v = cSubb.cross(hit_point.subVec3(b)).dot(normal);
+        const w = aSubc.cross(hit_point.subVec3(c)).dot(normal);
+
+        if (u < 0 or v < 0 or w < 0) {
+            return HitRecord.nil();
+        }
+
+        // map self.va.color, self.vb.color, self.vc.color
+        // to hit_point using barycentric coordinates
```
Comment (Miou-zora) [inline src/Triangle.zig:43]:
```suggestion
```

File: src/Triangle.zig [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        const material: Material = .{
+            .color = colorRGB,
+            .reflective = 0.5,
+            .specular = 0,
+        };
+        return HitRecord{
+            // .hit = true,
+            // .t = t,
+            // .hit_point = hit_point,
+            // .normal = normal,
```
Comment (Miou-zora) [inline src/Triangle.zig:56]:
```suggestion
```

File: src/main.zig (hunk 4 above)
Comment (Miou-zora) [inline src/main.zig:18]:
```suggestion
```

File: src/main.zig [hunk 11]
```diff
@@ -142,6 +142,13 @@ fn find_closest_intersection(scene: *Scene.Scene, ray: Ray, t_min: f32, t_max: f
                     }
                 }
             },
+            .triangle => |item| {
+                // std.debug.print("{}\n", .{item});
```
Comment (Miou-zora) [inline src/main.zig:146]:
```suggestion
```

File: src/main.zig [hunk 12]
```diff
@@ -223,6 +229,22 @@ pub fn main() !void {
...
+    //         .color = ColorRGB{ .r = 0, .g = 0, .b = 255 },
+    //         .position = Pt3{ .x = -2, .y = 3, .z = 3.5 },
+    //     },
+    //     .vc = .{
+    //         .color = ColorRGB{ .r = 0, .g = 255, .b = 0 },
+    //         .position = Pt3{ .x = 0.5, .y = -0.5, .z = 4 },
+    //     },
+    // } };
+    // try scene.objects.append(triangle);
+
```
Comment (Miou-zora) [inline src/main.zig:247]:
```suggestion
```

**Review by huntears — APPROVED** [review]

**Pull request merged**